// api/api.js
var API = function(){
  this.key = '?apiKey=5074a573e4b088be4c29efc4';
  this.baseUrl = 'https://api.mongolab.com/api/1';
  this.db = '/barbuddy';
}

API.prototype.verifyUser = function(username, password){
  this.getUser(username, function(users){
    if(users[0].password == password){
      alert('YAYYY logged in');
    } else {
      alert('Not logged in');
    }
  });
}

API.prototype.getUser = function(username, cb){
  this.getDocument('/users', '&q={"username":"' + username + '"}', cb)
}

API.prototype.addUser = function(username, password){
  var data = JSON.stringify({ 'username' : username, 'password' : password });
	this.addDocument('/users', data);
}

API.prototype.removeUser = function(username){
  this.removeDocument('/users', '{"username":"' + username + '"}');
}

API.prototype.getDocument = function(collection, query, cb){
  $.get(
     this.collectionUrl() + collection + this.key + query,
    function(){
      cb.apply(this, arguments);
    }
  );
}

API.prototype.addDocument = function(collection, data) {
  $.ajax({
    url: this.collectionUrl() + collection + this.key,
    data: data,
    type: "POST",
    contentType: "application/json"
  });
}

API.prototype.removeDocument = function(collection, query){
  $.ajax({
    url: this.collectionUrl() + collection + this.key + '&q=' + query,
    data: JSON.stringify([]),
    type: "PUT",
    contentType: "application/json"
  });
}

API.prototype.collectionUrl = function(){
  return this.baseUrl + '/databases' + this.db + '/collections';
}

module.exports = API;
