// api/api.js
var API = function(){
  this.key = '?apiKey=5074a573e4b088be4c29efc4';
  this.baseUrl = 'https://api.mongolab.com/api/1';
  this.db = '/barbuddy';
}

//=============================================================================
//User Calls

API.prototype.verifyUser = function(user, success, failure){
  this.getUser(user, function(users){
    if(users[0].password == user.password){
      success();
    } else {
      failure();
    }
  });
}

API.prototype.getUser = function(user, cb){
  this.getDocument('/users', user, cb)
}

API.prototype.addUser = function(user){
	this.addDocument('/users', user);
}

API.prototype.removeUser = function(user){
  this.removeDocument('/users', user);
}

//=============================================================================
//Bar Calls

API.prototype.getBar = function(username, cb){
  this.getDocument('/bars', '{"username":"' + username + '"}', cb);
}

API.prototype.addBar = function(bar){
  this.addDocument('/bars', bar);
}

API.prototype.removeBar = function(bar){
  this.removeDocument('/bars', bar);
}

API.prototype.replaceBar = function(bar){
  this.replaceDocument('/bars', bar, bar);
}

//=============================================================================
//Bottom Level Calls

API.prototype.getDocument = function(collection, queryObj, cb){
  $.get(
     this.collectionUrl() + collection + this.key + '&q=' + queryObj,
    function(){
      cb.apply(this, arguments);
    }
  );
}

API.prototype.addDocument = function(collection, data) {
  $.ajax({
    url: this.collectionUrl() + collection + this.key,
    data: JSON.stringify(data),
    type: "POST",
    contentType: "application/json"
  });
}

API.prototype.removeDocument = function(collection, queryObj){
  $.ajax({
    url: this.collectionUrl() + collection + this.key + '&q=' + queryObj,
    data: JSON.stringify([]),
    type: "PUT",
    contentType: "application/json"
  });
}

API.prototype.replaceDocument = function(collection, query, data){
  $.ajax({
    url: this.collectionUrl() + collection + this.key + '&q=' + query + '&u=true',
    data: JSON.stringify(data),
    type: "PUT",
    contentType: "application/json"
  });
}

API.prototype.collectionUrl = function(){
  return this.baseUrl + '/databases' + this.db + '/collections';
}

module.exports = API;
